False positive on `redundant_discardable_let` rule in SwiftUI `@ViewBuilder`

### New Issue Checklist

- [ x ] I've Updated SwiftLint to the latest version.
- [ x ] I've searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues).

### Bug Description

The rule `redundant_discardable_let` is falsely triggered inside SwiftUI `@ViewBuilder` implementation, although the option `ignore_swiftui_view_bodies` is set to `true`.
Using the `--fix` option for this violation will remove the `let`, and previews won't compile. The option is working properly inside SwiftUI `var body: some View` declarations.

```swift
// This triggers a violation:
@ViewBuilder
func viewBuilder() -> some View {
    let _ = "Test Builder"
    EmptyView()
}
```
Removing the `let` will break the `View` conformance. Adding explicit `return` to `EmptyView` will break the result builder `ViewBuilder`.

```bash
$ swiftlint --strict --progress

TestFile.swift:60:9: error: Redundant Discardable Let Violation: Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function (redundant_discardable_let)
```

### Environment

* SwiftLint version: 0.59.1
* Xcode version: Xcode 16.2, Build version 16C5032a
* Installation via Homebrew
* Configuration file:

```yml
# Default rule configuration
redundant_discardable_let:
    ignore_swiftui_view_bodies: true
```

No nested configurations.

Imho the exception for `var body: some View` should also apply to `@ViewBuilder` implementations since they follow the same rules as the `var body: some View` declarations.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -35,7 +35,9 @@
   or other error.  
   [Martin Redington](https://github.com/mildm8nnered)
   [#6052](https://github.com/realm/SwiftLint/issues/6052)
-* Fix false positives in `redundant_discardable_let` when `ignore_swiftui_view_bodies` is enabled.
+* Fix false positives of `redundant_discardable_let` rule in `@ViewBuilder` functions,
+  `#Preview` macro bodies and preview provides when `ignore_swiftui_view_bodies` is
+  enabled.  
   [kaseken](https://github.com/kaseken)
 
 ## 0.59.1: Crisp Spring Clean

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -29,16 +29,18 @@
 * Migrate `vertical_whitespace` rule from SourceKit to SwiftSyntax for improved performance.  
   [Matt Pennig](https://github.com/pennig)
 
+* Fix false positives of `redundant_discardable_let` rule in `@ViewBuilder` functions,
+  `#Preview` macro bodies and preview providers when `ignore_swiftui_view_bodies` is
+  enabled.  
+  [kaseken](https://github.com/kaseken)
+  [#6063](https://github.com/realm/SwiftLint/issues/6063)
+
 ### Bug Fixes
 
 * Improved error reporting when SwiftLint exits, because of an invalid configuration file
   or other error.  
   [Martin Redington](https://github.com/mildm8nnered)
   [#6052](https://github.com/realm/SwiftLint/issues/6052)
-* Fix false positives of `redundant_discardable_let` rule in `@ViewBuilder` functions,
-  `#Preview` macro bodies and preview provides when `ignore_swiftui_view_bodies` is
-  enabled.  
-  [kaseken](https://github.com/kaseken)
 
 ## 0.59.1: Crisp Spring Clean
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -35,6 +35,8 @@
   or other error.  
   [Martin Redington](https://github.com/mildm8nnered)
   [#6052](https://github.com/realm/SwiftLint/issues/6052)
+* Fix false positives in `redundant_discardable_let` when `ignore_swiftui_view_bodies` is enabled.
+  [kaseken](https://github.com/kaseken)
 
 ## 0.59.1: Crisp Spring Clean
 

--- a/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
@@ -153,9 +153,7 @@ private extension AccessorBlockSyntax {
         if let binding = parent?.as(PatternBindingSyntax.self),
            binding.pattern.as(IdentifierPatternSyntax.self)?.identifier.text == "body",
            let type = binding.typeAnnotation?.type.as(SomeOrAnyTypeSyntax.self) {
-            return type.someOrAnySpecifier.text == "some"
-                && type.constraint.as(IdentifierTypeSyntax.self)?.name.text == "View"
-                && binding.parent?.parent?.is(VariableDeclSyntax.self) == true
+            return type.isView && binding.parent?.parent?.is(VariableDeclSyntax.self) == true
         }
         return false
     }
@@ -171,8 +169,7 @@ private extension AccessorBlockSyntax {
             return false
         }
 
-        return type.someOrAnySpecifier.text == "some" &&
-            type.constraint.as(IdentifierTypeSyntax.self)?.name.text == "View"
+        return type.isView
     }
 }
 
@@ -187,18 +184,19 @@ private extension FunctionDeclSyntax {
         guard attributes.contains(attributeNamed: "ViewBuilder") else {
             return false
         }
-
-        guard let returnType = signature.returnClause?.type.as(SomeOrAnyTypeSyntax.self) else {
-            return false
-        }
-
-        return returnType.someOrAnySpecifier.text == "some" &&
-            returnType.constraint.as(IdentifierTypeSyntax.self)?.name.text == "View"
+        return signature.returnClause?.type.as(SomeOrAnyTypeSyntax.self)?.isView ?? false
     }
 }
 
 private extension ClosureExprSyntax {
     var isPreviewMacroBody: Bool {
         parent?.as(MacroExpansionExprSyntax.self)?.macroName.text == "Preview"
+    }
+}
+
+private extension SomeOrAnyTypeSyntax {
+    var isView: Bool {
+        someOrAnySpecifier.text == "some" &&
+            constraint.as(IdentifierTypeSyntax.self)?.name.text == "View"
     }
 }

--- a/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
@@ -54,19 +54,19 @@ struct RedundantDiscardableLetRule: Rule {
             Example("""
                 @ViewBuilder
                 func bar() -> some View {
-                    let _ = foo()
+                    ↓let _ = foo()
                     return Text("Hello, World!")
                 }
                 """),
             Example("""
                 #Preview {
-                    let _ = foo()
+                    ↓let _ = foo()
                     return Text("Hello, World!")
                 }
                 """),
             Example("""
                 static var previews: some View {
-                    let _ = foo()
+                    ↓let _ = foo()
                     Text("Hello, World!")
                 }
                 """),

--- a/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
@@ -86,6 +86,52 @@ struct RedundantDiscardableLetRule: Rule {
         corrections: [
             Example("↓let _ = foo()"): Example("_ = foo()"),
             Example("if _ = foo() { ↓let _ = bar() }"): Example("if _ = foo() { _ = bar() }"),
+            Example("""
+                var body: some View {
+                    ↓let _ = foo()
+                    Text("Hello, World!")
+                }
+                """): Example("""
+                    var body: some View {
+                        _ = foo()
+                        Text("Hello, World!")
+                    }
+                    """),
+            Example("""
+                @ViewBuilder
+                func bar() -> some View {
+                    ↓let _ = foo()
+                    return Text("Hello, World!")
+                }
+                """): Example("""
+                    @ViewBuilder
+                    func bar() -> some View {
+                        _ = foo()
+                        return Text("Hello, World!")
+                    }
+                    """),
+            Example("""
+                #Preview {
+                    ↓let _ = foo()
+                    return Text("Hello, World!")
+                }
+                """): Example("""
+                    #Preview {
+                        _ = foo()
+                        return Text("Hello, World!")
+                    }
+                    """),
+            Example("""
+                static var previews: some View {
+                    ↓let _ = foo()
+                    Text("Hello, World!")
+                }
+                """): Example("""
+                    static var previews: some View {
+                        _ = foo()
+                        Text("Hello, World!")
+                    }
+                    """),
         ]
     )
 }

--- a/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
@@ -70,6 +70,18 @@ struct RedundantDiscardableLetRule: Rule {
                     Text("Hello, World!")
                 }
                 """),
+            Example("""
+                var notBody: some View {
+                    ↓let _ = foo()
+                    Text("Hello, World!")
+                }
+                """, configuration: ["ignore_swiftui_view_bodies": true], excludeFromDocumentation: true),
+            Example("""
+                var body: some NotView {
+                    ↓let _ = foo()
+                    Text("Hello, World!")
+                }
+                """, configuration: ["ignore_swiftui_view_bodies": true], excludeFromDocumentation: true),
         ],
         corrections: [
             Example("↓let _ = foo()"): Example("_ = foo()"),

--- a/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
@@ -152,21 +152,15 @@ private extension AccessorBlockSyntax {
         guard let binding = parent?.as(PatternBindingSyntax.self),
               binding.pattern.as(IdentifierPatternSyntax.self)?.identifier.text == "previews",
               let bindingList = binding.parent?.as(PatternBindingListSyntax.self),
-              let variableDecl = bindingList.parent?.as(VariableDeclSyntax.self) else {
+              let variableDecl = bindingList.parent?.as(VariableDeclSyntax.self),
+              variableDecl.modifiers.contains(keyword: .static),
+              variableDecl.bindingSpecifier.tokenKind == .keyword(.var),
+              let type = binding.typeAnnotation?.type.as(SomeOrAnyTypeSyntax.self) else {
             return false
         }
 
-        guard variableDecl.modifiers.contains(keyword: .static) &&
-                variableDecl.bindingSpecifier.tokenKind == .keyword(.var) else {
-            return false
-        }
-
-        if let type = binding.typeAnnotation?.type.as(SomeOrAnyTypeSyntax.self) {
-            return type.someOrAnySpecifier.text == "some" &&
-                type.constraint.as(IdentifierTypeSyntax.self)?.name.text == "View"
-        }
-
-        return false
+        return type.someOrAnySpecifier.text == "some" &&
+            type.constraint.as(IdentifierTypeSyntax.self)?.name.text == "View"
     }
 }
 

--- a/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
@@ -199,10 +199,6 @@ private extension FunctionDeclSyntax {
 
 private extension ClosureExprSyntax {
     var isPreviewMacroBody: Bool {
-        if let macroExpansionExpr = parent?.as(MacroExpansionExprSyntax.self),
-           macroExpansionExpr.macroName.text == "Preview" {
-            return true
-        }
-        return false
+        parent?.as(MacroExpansionExprSyntax.self)?.macroName.text == "Preview"
     }
 }

--- a/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
@@ -98,19 +98,6 @@ struct RedundantDiscardableLetRule: Rule {
                     }
                     """),
             Example("""
-                @ViewBuilder
-                func bar() -> some View {
-                    ↓let _ = foo()
-                    return Text("Hello, World!")
-                }
-                """): Example("""
-                    @ViewBuilder
-                    func bar() -> some View {
-                        _ = foo()
-                        return Text("Hello, World!")
-                    }
-                    """),
-            Example("""
                 #Preview {
                     ↓let _ = foo()
                     return Text("Hello, World!")
@@ -122,14 +109,14 @@ struct RedundantDiscardableLetRule: Rule {
                     }
                     """),
             Example("""
-                static var previews: some View {
-                    ↓let _ = foo()
-                    Text("Hello, World!")
+                var body: some View {
+                    let _ = foo()
+                    return Text("Hello, World!")
                 }
-                """): Example("""
-                    static var previews: some View {
-                        _ = foo()
-                        Text("Hello, World!")
+                """, configuration: ["ignore_swiftui_view_bodies": true]): Example("""
+                    var body: some View {
+                        let _ = foo()
+                        return Text("Hello, World!")
                     }
                     """),
         ]

--- a/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
@@ -175,16 +175,11 @@ private extension AccessorBlockSyntax {
 
 private extension CodeBlockSyntax {
     var isViewBuilderFunctionBody: Bool {
-        parent?.as(FunctionDeclSyntax.self)?.isViewBuilderFunction == true
-    }
-}
-
-private extension FunctionDeclSyntax {
-    var isViewBuilderFunction: Bool {
-        guard attributes.contains(attributeNamed: "ViewBuilder") else {
+        guard let functionDecl = parent?.as(FunctionDeclSyntax.self),
+              functionDecl.attributes.contains(attributeNamed: "ViewBuilder") else {
             return false
         }
-        return signature.returnClause?.type.as(SomeOrAnyTypeSyntax.self)?.isView ?? false
+        return functionDecl.signature.returnClause?.type.as(SomeOrAnyTypeSyntax.self)?.isView ?? false
     }
 }
 

--- a/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
@@ -22,6 +22,25 @@ struct RedundantDiscardableLetRule: Rule {
                     return Text("Hello, World!")
                 }
                 """, configuration: ["ignore_swiftui_view_bodies": true]),
+            Example("""
+                @ViewBuilder
+                func bar() -> some View {
+                    let _ = foo()
+                    Text("Hello, World!")
+                }
+                """, configuration: ["ignore_swiftui_view_bodies": true]),
+            Example("""
+                #Preview {
+                    let _ = foo()
+                    Text("Hello, World!")
+                }
+                """, configuration: ["ignore_swiftui_view_bodies": true]),
+            Example("""
+                static var previews: some View {
+                    let _ = foo()
+                    Text("Hello, World!")
+                }
+                """, configuration: ["ignore_swiftui_view_bodies": true]),
         ],
         triggeringExamples: [
             Example("↓let _ = foo()"),
@@ -29,6 +48,25 @@ struct RedundantDiscardableLetRule: Rule {
             Example("""
                 var body: some View {
                     ↓let _ = foo()
+                    Text("Hello, World!")
+                }
+                """),
+            Example("""
+                @ViewBuilder
+                func bar() -> some View {
+                    let _ = foo()
+                    return Text("Hello, World!")
+                }
+                """),
+            Example("""
+                #Preview {
+                    let _ = foo()
+                    return Text("Hello, World!")
+                }
+                """),
+            Example("""
+                static var previews: some View {
+                    let _ = foo()
                     Text("Hello, World!")
                 }
                 """),
@@ -50,7 +88,7 @@ private extension RedundantDiscardableLetRule {
         private var codeBlockScopes = Stack<CodeBlockKind>()
 
         override func visit(_ node: AccessorBlockSyntax) -> SyntaxVisitorContinueKind {
-            codeBlockScopes.push(node.isViewBody ? .view : .normal)
+            codeBlockScopes.push(node.isViewBody || node.isPreviewProviderBody ? .view : .normal)
             return .visitChildren
         }
 
@@ -58,12 +96,21 @@ private extension RedundantDiscardableLetRule {
             codeBlockScopes.pop()
         }
 
-        override func visit(_: CodeBlockSyntax) -> SyntaxVisitorContinueKind {
-            codeBlockScopes.push(.normal)
+        override func visit(_ node: CodeBlockSyntax) -> SyntaxVisitorContinueKind {
+            codeBlockScopes.push(node.isViewBuilderFunctionBody ? .view : .normal)
             return .visitChildren
         }
 
         override func visitPost(_: CodeBlockSyntax) {
+            codeBlockScopes.pop()
+        }
+
+        override func visit(_ node: ClosureExprSyntax) -> SyntaxVisitorContinueKind {
+            codeBlockScopes.push(node.isPreviewMacroBody ? .view : .normal)
+            return .visitChildren
+        }
+
+        override func visitPost(_: ClosureExprSyntax) {
             codeBlockScopes.pop()
         }
 
@@ -97,6 +144,58 @@ private extension AccessorBlockSyntax {
             return type.someOrAnySpecifier.text == "some"
                 && type.constraint.as(IdentifierTypeSyntax.self)?.name.text == "View"
                 && binding.parent?.parent?.is(VariableDeclSyntax.self) == true
+        }
+        return false
+    }
+
+    var isPreviewProviderBody: Bool {
+        guard let binding = parent?.as(PatternBindingSyntax.self),
+              binding.pattern.as(IdentifierPatternSyntax.self)?.identifier.text == "previews",
+              let bindingList = binding.parent?.as(PatternBindingListSyntax.self),
+              let variableDecl = bindingList.parent?.as(VariableDeclSyntax.self) else {
+            return false
+        }
+
+        guard variableDecl.modifiers.contains(keyword: .static) &&
+                variableDecl.bindingSpecifier.tokenKind == .keyword(.var) else {
+            return false
+        }
+
+        if let type = binding.typeAnnotation?.type.as(SomeOrAnyTypeSyntax.self) {
+            return type.someOrAnySpecifier.text == "some" &&
+                type.constraint.as(IdentifierTypeSyntax.self)?.name.text == "View"
+        }
+
+        return false
+    }
+}
+
+private extension CodeBlockSyntax {
+    var isViewBuilderFunctionBody: Bool {
+        parent?.as(FunctionDeclSyntax.self)?.isViewBuilderFunction == true
+    }
+}
+
+private extension FunctionDeclSyntax {
+    var isViewBuilderFunction: Bool {
+        guard attributes.contains(attributeNamed: "ViewBuilder") else {
+            return false
+        }
+
+        guard let returnType = signature.returnClause?.type.as(SomeOrAnyTypeSyntax.self) else {
+            return false
+        }
+
+        return returnType.someOrAnySpecifier.text == "some" &&
+            returnType.constraint.as(IdentifierTypeSyntax.self)?.name.text == "View"
+    }
+}
+
+private extension ClosureExprSyntax {
+    var isPreviewMacroBody: Bool {
+        if let macroExpansionExpr = parent?.as(MacroExpansionExprSyntax.self),
+           macroExpansionExpr.macroName.text == "Preview" {
+            return true
         }
         return false
     }

